Extract mappings to their  own DTOs

greg0ire opened the pull request:
An experiment, not sure if I will manage to do it, but at least it will allow us to port some changes to 2.x:
- for instance, everywhere `FieldMapping` the DTO is used, `FieldMapping` the array-shape should be used
- we can add missing fields to `FieldMapping` (for instance, `version`, or `default`)

## Todo

- [x] field mapping
- [x] association mapping
- [x] embedded classes mapping
- [x] discriminator column mapping
- [x] avoid repetition (array access implementation,  common fields)
- [x] use different DTOs for different association types
- [x] use different DTOs for different sides of the association
- [x] make classes that should never be instantiated abstract
- [x] Fix Psalm issues
- [x] Fix PHPStan issues
- [x] Implement `__sleep()`
- [x] Avoid redundant `isCascade*` properties

mpdude commented:
What is the supposed migration path for this? 

In 2.x, the `ClassMetadataInfo::$fieldMappings` is a public property, marked as `READ-ONLY`, but not internal. To my understanding, that means it is allowed to read and work with them.

I don't know about the deprecation/transition rules for this project. But changing this on the 3.0 branch means things will break during the upgrade, and there is no "smooth" transition (e. g. by getting deprecation warnings in 2.x, which can be fixed, then the 3.0 upgrade can be made).

I generally support the idea and think it's a good thing design-wise.

Can we make that more user-friendly, by
* adding the DTO in 2.x already, 
* implementing `ArrayAccess`, make things look like before
* trigger a deprecation when the array access is actually used
* drop the array access in 3.0?

greg0ire commented:
I'd actually add the array access on 3.0,much like what I did recently for the Token class in doctrine/lexer, that way the type does not change from array to DTO in a minor. What we might do to ease the transition could be to introduce a factory that would produce arrays with nullable properties instead of optional properties, to make the array closer to the final dto

greg0ire committed:
Extract FieldMapping in its own DTO

greg0ire commented on lib/Doctrine/ORM/Mapping/FieldMapping.php:
```diff
@@ -0,0 +1,117 @@
+    /** @var int|null The database length of the column. Optional. Default value taken from the type. */
+    public int|null $length = null;
+    /**
+     * @var bool|null Marks the field as the primary key of the entity. Multiple
```
While trying to fix the build, I reverted to `bool|null` here, to be closer to the previous situation. Now that the build is green, I can do a separate commit to make boolean field not-nullable, and default to `false`.

greg0ire commented on lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php:
```diff
@@ -328,15 +328,16 @@ private function getShortName(string $className): string
         foreach ($parentClass->fieldMappings as $mapping) {
-            if (! isset($mapping['inherited']) && ! $parentClass->isMappedSuperclass) {
-                $mapping['inherited'] = $parentClass->name;
+            $subClassMapping = clone $mapping;
```
Lost a huge amount of time on this, but yeah, we have to clone here now.

mpdude replied:
#10397 changed that area, did you include/merge that already?

why can’t you just update the value?

greg0ire replied:
I need to update the child value without updating the parent value. When using an array, a copy is made. When using objects, we're altering the parent objects.

We do need to do a merge up.

mpdude replied:
Would it be better to do the clone up in `doLoadMetadata` or so, in a more central location?

greg0ire replied:
Merge up done in https://github.com/doctrine/orm/pull/10421

About `doLoadMetadata`, I'm not sure… I mean there's one clone per mapping to do, so it's not like you could pass clones to `addInheritedFields` or something :thinking:

greg0ire commented:
Did a new push, comes with static analysis issues that I should be able to solve once I have converted every argument to addMappingInheritanceInformation to a DTO.

greg0ire commented:
Pushed a new version with an association mapping DTO, hopefully that was the hardest part of the job.

greg0ire committed:
Extract discriminator column mapping into its own DTO

greg0ire committed:
Extract embedded class mapping into its own DTO

greg0ire commented:
Just found out that I'm doing the opposite of what's been done 13 years ago in 8d3e0e61eaeaaf8753838a8580b2ee36dbd8f34f :sweat_smile:

mpdude commented:
Maybe performance? Arrays might be faster to access than methods to call?

greg0ire commented:
Discussed it internally, and you are right… well, you were. It was true in PHP 5.3, but no longer is.

mpdude commented:
Before you put too much work into it, are you sure there is a smooth migration path?

greg0ire commented:
I think the `ArrayAccess` implementation is enough. All the arrays I'm changing are object-like arrays, so they are very unlikely to be used with anything else than the `[]` notation. Also, I experimented with this on `doctrine/lexer` and got no complaints.

greg0ire committed:
Separate classes

greg0ire committed:
leverage To*AssociationMapping classes

greg0ire committed:
Introduce abstractions for owning side

greg0ire committed:
Extract AssociationMapping in its own DTO

greg0ire commented on lib/Doctrine/ORM/Mapping/AssociationMapping.php:
```diff
@@ -0,0 +1,226 @@
+    public function isManyToManyOwningSide(): bool
+    {
+        return $this instanceof ManyToManyOwningSideMapping;
+    }
```
@mpdude doing `instanceof` checks got old real fast, so I introduced the 3 methods above, and created https://github.com/phpstan/phpstan/issues/8904

greg0ire committed:
Introduce AssociationMapping::isToOne()

greg0ire committed:
Remove redundant AssociationMapping::$type

greg0ire committed:
Make abstract classes that should be

greg0ire committed:
Move joinColumns to where they belong

greg0ire committed:
Use psalm-param to fix cs issue

greg0ire committed:
Use more precise/accurate docblocks

greg0ire committed:
Move validation logic to dtos

greg0ire commented:
@phansys @franmomu hi! In this PR, I change arrays into DTOs, and I've been told there might be places in `gedmo/doctrine-extensions` where extra properties are written to these arrays. You seem to be active on those repos, so if this rings a bell, please point me to where this happens, so that I can come up with a plan to support this.

greg0ire committed:
Extract array access implementation into a trait

franmomu commented:
> @phansys @franmomu hi! In this PR, I change arrays into DTOs, and I've been told there might be places in `gedmo/doctrine-extensions` where extra properties are written to these arrays. You seem to be active on those repos, so if this rings a bell, please point me to where this happens, so that I can come up with a plan to support this.

hey @greg0ire, off the top of my head I don't remember modifying these arrays directly, there are usually calls through the `UoW`, but I have to take a look. What I think we might have are calls to `array_key_exists` to these arrays [which will fail](https://3v4l.org/rGDm9), but we can change that beforehand.

greg0ire commented:
Thanks for looking into this. We can always add support for extra properties later if they are a must anyway.

greg0ire committed:
switch to static

greg0ire committed:
Remove unused baseline entries

greg0ire committed:
Baseline remaining issues

greg0ire commented:
I removed the item "implement `__sleep()`", I think it would be best done later with a performance benchmark to determine whether it's worth it or not. Let's mark this as ready for review then.

derrabus commented on lib/Doctrine/ORM/Mapping/ArrayAccessImplementation.php:
```diff
@@ -0,0 +1,35 @@
+trait ArrayAccessImplementation
+{
+    /** @param array-key $offset */
+    public function offsetExists($offset): bool
```
```suggestion
    public function offsetExists(mixed $offset): bool
```

derrabus commented on lib/Doctrine/ORM/Mapping/ArrayAccessImplementation.php:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    /** @param array-key $offset */
+    public function offsetGet($offset): mixed
```
```suggestion
    public function offsetGet(mixed $offset): mixed
```

derrabus commented on lib/Doctrine/ORM/Mapping/ArrayAccessImplementation.php:
```diff
@@ -0,0 +1,35 @@
+     * @param array-key $offset
+     * @param mixed     $value
+     */
+    public function offsetSet($offset, $value): void
```
```suggestion
    public function offsetSet(mixed $offset, mixed $value): void
```

derrabus commented on lib/Doctrine/ORM/Mapping/ArrayAccessImplementation.php:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    /** @param array-key $offset */
+    public function offsetUnset($offset): void
```
```suggestion
    public function offsetUnset(mixed $offset): void
```

derrabus commented on lib/Doctrine/ORM/Mapping/ArrayAccessImplementation.php:
```diff
@@ -0,0 +1,35 @@
+
+namespace Doctrine\ORM\Mapping;
+
+trait ArrayAccessImplementation
```
`@internal` please!!! 🙀

derrabus commented on lib/Doctrine/ORM/Mapping/ArrayAccessImplementation.php:
```diff
@@ -0,0 +1,35 @@
+
+trait ArrayAccessImplementation
+{
+    /** @param array-key $offset */
```
All of the consumers of this trait use string offsets, so we might as well…

```suggestion
    /** @param string $offset */
```

Apart from that, PHPStan and Psalm should support generic traits.

derrabus commented on lib/Doctrine/ORM/Mapping/AssociationMapping.php:
```diff
@@ -0,0 +1,240 @@
+    public function isOwningSide(): bool
+    {
+        return $this instanceof AssociationOwningSideMapping;
+    }
```
Maybe I'm getting old, but wouldn't the saner approach be…

```suggestion
    public function isOwningSide(): bool
    {
        return false;
    }
```

… and overriding this method in `AssociationOwningSideMapping`?

greg0ire replied:
It doesn't look saner to me. When stumbling upon the code in your proposition, one has might think "what's going on here? Ah, maybe that method is overridden somewhere (but where?). With the approach I picked, you have everything you need to understand this in the same place.

derrabus replied:
> Ah, maybe that method is overridden somewhere

Yeah, that's the general idea of abstract classes. 😅 

It's a bit weird that the abstraction knows details about its implementations. But all right, let's roll with it for now.

greg0ire replied:
> It's a bit weird that the abstraction knows details about its implementations

I'd like to make this a bit more like a sealed class (https://wiki.php.net/rfc/sealed_classes), but I didn't know how. But yeah, that's indeed the concept here.

derrabus commented on lib/Doctrine/ORM/Mapping/AssociationMapping.php:
```diff
@@ -0,0 +1,240 @@
+        return $this instanceof ManyToManyOwningSideMapping;
+    }
+
+    public function type(): int
+    {
+        return match (true) {
+            $this instanceof OneToOneAssociationMapping => ClassMetadata::ONE_TO_ONE,
+            $this instanceof OneToManyAssociationMapping => ClassMetadata::ONE_TO_MANY,
+            $this instanceof ManyToOneAssociationMapping => ClassMetadata::MANY_TO_ONE,
+            $this instanceof ManyToManyAssociationMapping => ClassMetadata::MANY_TO_MANY,
+            default => throw new Exception('Cannot determine type for ' . $this::class),
+        };
+    }
```
We could make a `public readonly int $type` that is initialized by the constructor using this `match` statement. This would remove an edge-case from your `offsetGet`/`offsetSet` and `toArray` implementations.

greg0ire replied:
I feel uneasy duplicating information contained in the type itself as a property.

derrabus replied:
Understood. I was trying to eliminate that special treatment from those three methods, but maybe I was trying to hard. A method is probably the better approach.

derrabus commented on lib/Doctrine/ORM/Mapping/AssociationMapping.php:
```diff
@@ -0,0 +1,240 @@
+        };
+    }
+
+    public function offsetGet($offset): mixed
```
You're overriding methods that are provided by the `ArrayAccessImplementation` trait already. You do this to cover two edge cases. The methods `offsetExists()` and `offsetUnset()` are currently broken because those edge cases are not taken into account.

Suggestion:

* Either don't use the trait in this class and implement all `ArrayAccess` methods in a consistent manner,
* or eliminate the edge-cases, or at least make them work without having to override the trait.

derrabus commented on lib/Doctrine/ORM/Mapping/AssociationMapping.php:
```diff
@@ -0,0 +1,240 @@
+            return $this->type();
+        }
+
+        return $this->$offset;
```
Y U NO `match($offset)`?

derrabus commented on lib/Doctrine/ORM/Mapping/ArrayAccessImplementation.php:
```diff
@@ -0,0 +1,35 @@
+    /** @param array-key $offset */
+    public function offsetGet($offset): mixed
+    {
+        return $this->$offset;
```
Should we check if the property exists? Or do we want to allow arbitrary properties? In the latter case, we should put `#[AllowDynamicProperties]` on all implementations.

greg0ire replied:
I added the check, and at some point it didn't work, so I removed it. I just tried adding it back, and it the test suite passes. I'll push to see if it works with no skipped tests.

greg0ire replied:
All checks pass!

derrabus commented on lib/Doctrine/ORM/Mapping/AssociationMapping.php:
```diff
@@ -0,0 +1,240 @@
+    public function offsetSet($offset, $value): void
+    {
+        if ($offset === 'joinTable') {
+            assert($this instanceof ManyToManyAssociationMapping);
```
We should raise a proper exception instead.

greg0ire replied:
Why? I think we should avoid new exceptions for temporary code such as this. It has no equivalent for properties.

derrabus replied:
So the plan is to remove the `ArrayAccess` implementations before the release? In that case, I'm fine with the assert. Otherwise we need to assume that the array access is used downstream and this assert is basically input validation which is not what assert should be used for.

greg0ire replied:
No, the plan is to stop using array access internally before the release, to deprecate it in 3.1, and to remove it in 4.0.

I'll remove this assert, is was just a way for me to make sure I was doing this refactoring properly (checking my assumptions).

derrabus commented on lib/Doctrine/ORM/Mapping/AssociationMapping.php:
```diff
@@ -0,0 +1,240 @@
+     *                                   source entity.
+     */
+    final public function __construct(
+        public string $fieldName,
+        public string $sourceEntity,
+        public string $targetEntity,
```
Can we make those `readonly`? I don't think we want to mutate them.

greg0ire replied:
Not `$sourceEntity` according to the test suite, because of
https://github.com/doctrine/orm/blob/13d86957808f3fab3dfb9c681c59984a2bb4e4dd/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L444

It works for the other ones though.

derrabus commented on lib/Doctrine/ORM/Mapping/DiscriminatorColumnMapping.php:
```diff
@@ -0,0 +1,57 @@
+                continue;
+            }
+
+            if (property_exists($mapping, $key)) {
+                $mapping->$key = $value;
+            }
```
In other classes, you throw for invalid properties while you silently ignore them here. Is that on purpose?

greg0ire replied:
IIRC I started added that check in other classes after this one, I must have forgotten to update it.

derrabus commented on lib/Doctrine/ORM/Mapping/AssociationMapping.php:
```diff
@@ -0,0 +1,240 @@
+use function property_exists;
+
+/** @template-implements ArrayAccess<string, mixed> */
+abstract class AssociationMapping implements ArrayAccess
```
Can we `final`ize all public methods in this class that are not overridden by child classes?

derrabus commented on lib/Doctrine/ORM/Mapping/JoinColumnData.php:
```diff
@@ -0,0 +1,46 @@
+use function property_exists;
+
+/** @template-implements ArrayAccess<string, mixed> */
+final class JoinColumnData implements ArrayAccess
```
Data is very generic and doesn't seem like a good fit. `JoinColumnMapping`? `JoinColumnDefinition`? `JoinColumn`?

greg0ire replied:
`JoinColumnMapping` 👍

derrabus commented on lib/Doctrine/ORM/Mapping/JoinColumnData.php:
```diff
@@ -0,0 +1,46 @@
+    /** @var array<string, mixed> */
+    public array|null $options = null;
+
+    public function __construct()
+    {
+    }
```
Can be removed.

derrabus commented on lib/Doctrine/ORM/Mapping/ManyToManyOwningSideMapping.php:
```diff
@@ -0,0 +1,131 @@
+
+        // owning side MUST have a join table
+        if (! isset($mapping['joinTable']['name'])) {
+            if (! isset($mapping->joinTable)) {
+                $mapping->joinTable = new JoinTableMapping();
+            }
```
```suggestion
            $mapping->joinTable ??= new JoinTableMapping();
```

derrabus commented on lib/Doctrine/ORM/Mapping/JoinTableMapping.php:
```diff
@@ -0,0 +1,87 @@
+
+    public bool|null $quoted = null;
+
+    /** @var list<JoinColumnData> */
+    public array|null $joinColumns = null;
```
The native type allows `null` while the `@var` type doesn't. This should be fixed. Are `null` and `[]` actually different cases? If not, I suggest to remove the nullability here.

greg0ire replied:
`[]` causes an exception from DBAL on this piece of code:

```
        if (count($foreignKey->getLocalColumns()) === 0) {
            throw new InvalidArgumentException("Incomplete definition. 'local' required.");
        }
```

So I guess they are different cases, and `[]` is not supposed to ever happen.

derrabus replied:
But is `null` supposed to happen? A join table without columns doesn't sound right to me either.

greg0ire replied:
Disallowing `null` seems OK according to the test suite, defaulting to `[]` isn't, though… so I guess it's not `null`, but uninitialized? I'm confused now…

derrabus replied:
> so I guess it's not `null`, but uninitialized? I'm confused now…

What we're looking for is a representation for an incomplete `JoinTableMapping`. You construct that object and then you add columns to it. For that matter, the empty array seems like a good initial value. `null` is just an unnecessary edge case that static analysis will always force us to cover.

greg0ire replied:
I took a closer look at the failing test, it fails because of what's in `ManyToManyOwningSideMapping`:

```
        if (! isset($mapping->joinTable->joinColumns)) {
            $mapping->joinTable->joinColumns = [
                JoinColumnMapping::fromMappingArray([
                    'name' => $namingStrategy->joinKeyColumnName($mapping->sourceEntity, $selfReferencingEntityWithoutJoinColumns ? 'source' : null),
                    'referencedColumnName' => $namingStrategy->referenceColumnName(),
                    'onDelete' => 'CASCADE',
                ]),
            ];
        }
```

Using `=== []` instead of `isset()` fixes this particular issue, I'm looking at the next one.

greg0ire replied:
fixed :)

derrabus commented on lib/Doctrine/ORM/Mapping/JoinTableMapping.php:
```diff
@@ -0,0 +1,87 @@
+    /** @var list<JoinColumnData> */
+    public array|null $joinColumns = null;
+
+    /** @var list<JoinColumnData> */
+    public array|null $inverseJoinColumns = null;
```
Same.

derrabus commented on lib/Doctrine/ORM/Mapping/ManyToManyAssociationMapping.php:
```diff
@@ -0,0 +1,23 @@
+    {
+        $mapping = parent::fromMappingArray($mappingArray);
+
+        $mapping['orphanRemoval'] = isset($mapping['orphanRemoval']) && $mapping['orphanRemoval'];
```
This line confuses me. Why are you mapping this property to itself?

greg0ire replied:
I'm just moving code around here:

https://github.com/doctrine/orm/blob/13d86957808f3fab3dfb9c681c59984a2bb4e4dd/lib/Doctrine/ORM/Mapping/ClassMetadata.php#L1685

This ensures the property ends up being a boolean, treating `null` as `false`. I will remove this line and default to `false` instead.

derrabus commented on lib/Doctrine/ORM/Mapping/ManyToManyOwningSideMapping.php:
```diff
@@ -0,0 +1,131 @@
+                $mapping->joinTable = new JoinTableMapping();
+            }
+
+            $mapping->joinTable['name'] = $namingStrategy->joinTableName(
```
I noticed that you're switching between array access and properties within the new DTOs. I understand that keeping array access keeps the diff small for old code, but inside the new classes, I believe using properties consistently would improve the readability. Especially in methods where we map from arrays to DTOs and vice versa, using properties would make it clearer what is the array and what the DTO.

In this very line, you're even using both methods in one expression. 😵

derrabus reviewed (changes requested):
🤯

greg0ire committed:
Make readonly properties that can be

greg0ire committed:
Finalize methods that can be finalized

greg0ire committed:
Rename JoinColumnData to JoinColumnMapping

greg0ire committed:
Remove unused constructor

greg0ire committed:
Use null coalesce assignment operator

greg0ire committed:
Document properties as nullable

greg0ire committed:
Default orphan removal to false

greg0ire committed:
Avoid array access in new files

derrabus commented on lib/Doctrine/ORM/Mapping/DiscriminatorColumnMapping.php:
```diff
@@ -0,0 +1,60 @@
+{
+    use ArrayAccessImplementation;
+
+    /** @var int|null The database length of the column. Optional. Default value taken from the type. */
```
```suggestion
    /** The database length of the column. Optional. Default value taken from the type. */
```

greg0ire committed:
Avoid array access trait when overridden

derrabus commented on lib/Doctrine/ORM/Mapping/EmbeddedClassMapping.php:
```diff
@@ -0,0 +1,66 @@
+     * mapped superclasses are <em>not</em> considered 'inherited' in the
+     * nearest entity subclasses.
+     *
+     * @var class-string
```
```suggestion
     * @var class-string|null
```

derrabus commented on lib/Doctrine/ORM/Mapping/EmbeddedClassMapping.php:
```diff
@@ -0,0 +1,66 @@
+     * topmost non-transient class that contains mapping information for this
+     * field.
+     *
+     * @var class-string
```
```suggestion
     * @var class-string|null
```

greg0ire committed:
Remove nullability for join columns

beberlei reviewed (commented):
I didnt do a full review, but I like the parts i saw. If you are confident about this change you have my +1.

The only thing insm wondering about are  some of the duplications, for example the isCascade... - specifically I am wondering if all these new mapping claasea shoulf have code that omits the default values to end up in the serialized payloads for metadata cache, like its done in ClassMetadata

greg0ire commented:
You mean `__sleep()` methods? I see there were such methods in 8d3e0e61eaeaaf8753838a8580b2ee36dbd8f34f, before switching to arrays, so I can definitely look into doing something similar.

greg0ire committed:
Serialize only what's necessary

greg0ire commented:
> some of the duplications, for example the isCascade

Oh right I didn't spot that, but it looks like the properties `AssociationMapping::$isCascade*` could become methods, based on `AssociationMapping::$cascade`.

greg0ire committed:
Stop synchronizing properties

Instead of synchronizing isCascade* with cascade, we dynamically compute
these methods.

greg0ire commented:
@beberlei my 2 last commits should address your concerns :)

greg0ire commented on lib/Doctrine/ORM/Mapping/DiscriminatorColumnMapping.php:
```diff
@@ -0,0 +1,74 @@
+
+/**
+ * @template-implements ArrayAccess<string, mixed>
+ * @psalm-type DiscriminatorColumnMappingType = 'boolean'|'array'|'object'|'datetime'|'time'|'date'|'string'
```
This is wrong. These are values that `$type` can _not_ take.

greg0ire replied:
Fixed

greg0ire committed:
Narrow type down

greg0ire committed:
Remove reference access

Now that we deal with objects, everything is simpler.

greg0ire committed:
Make referencedColumnName mandatory

When switching to an object API, it becomes obvious that it should be
mandatory. You have to assert it is not null everywhere.

SenseException commented on lib/Doctrine/ORM/Cache/DefaultCache.php:
```diff
@@ -157,7 +158,7 @@ public function evictCollectionRegions(): void
         foreach ($metadatas as $metadata) {
             foreach ($metadata->associationMappings as $association) {
-                if (! $association['type'] & ClassMetadata::TO_MANY) {
+                if (! $association instanceof ToManyAssociationMapping) {
```
If those are all `AssociationMapping`, wouldn't the use of a method be an option instead of an `instanceof` check? You've used methods at other places too.

greg0ire replied:
You're right, I will add such a method.

SenseException reviewed (commented):
I haven't done the full review yet. The amount of changes slows my progress, so maybe I'll do it in parts.

greg0ire commented:
Closing in favor of smaller PRs, the first of which will be https://github.com/doctrine/orm/pull/10607
